File: src/org/elixir_lang/psi/scope/module/MultiResolve.java
package org.elixir_lang.psi.scope.module;

import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndex;
import com.intellij.psi.util.PsiTreeUtil;
import org.elixir_lang.psi.NamedElement;
import org.elixir_lang.psi.call.Named;
import org.elixir_lang.psi.scope.Module;
import org.elixir_lang.psi.stub.index.AllName;
import org.elixir_lang.reference.module.UnaliasedName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import static org.elixir_lang.Module.concat;
import static org.elixir_lang.Module.split;
import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE;

public class MultiResolve extends Module {
    /*
     * Public Static Methods
     */

    @Nullable
    public static List<ResolveResult> resolveResultList(@NotNull String name,
                                                        boolean incompleteCode,
                                                        @NotNull PsiElement entrance,
                                                        @NotNull PsiElement maxScope) {
      return resolveResultList(name, incompleteCode, entrance, maxScope, ResolveState.initial());
    }

    /*
     * Private Static Methods
     */

    @NotNull
    private static Collection<NamedElement> indexedNamedElements(@NotNull PsiNamedElement match,
                                                                 @NotNull String unaliasedName) {
        Project project = match.getProject();
        return StubIndex.getElements(
                AllName.KEY,
                unaliasedName,
                project,
                GlobalSearchScope.allScope(project),
                NamedElement.class
        );
    }

    @Nullable
    private static List<ResolveResult> resolveResultList(@NotNull String name,
                                                         boolean incompleteCode,
                                                         @NotNull PsiElement entrance,
                                                         @NotNull PsiElement maxScope,
                                                         @NotNull ResolveState state) {
        MultiResolve multiResolve = new MultiResolve(name, incompleteCode);
        PsiTreeUtil.treeWalkUp(
                multiResolve,
                entrance,
                maxScope,
                state.put(ENTRANCE, entrance)
        );
        return multiResolve.getResolveResultList();
    }

    @NotNull
    private static String unaliasedName(@NotNull PsiNamedElement match, @NotNull List<String> namePartList) {
        String matchUnaliasedName = UnaliasedName.unaliasedName(match);

        List<String> unaliasedNamePartList = new ArrayList<String>(namePartList.size());
        unaliasedNamePartList.add(matchUnaliasedName);

        for (int i = 1; i < namePartList.size(); i++) {
            unaliasedNamePartList.add(namePartList.get(i));
        }

        return concat(unaliasedNamePartList);
    }

    /*
     * Fields
     */

    @NotNull
    private final String name;
    private final boolean incompleteCode;
    @Nullable
    private List<ResolveResult> resolveResultList = null;

    /*
     * Constructors
     */

    MultiResolve(@NotNull String name, boolean incompleteCode) {
        this.incompleteCode = incompleteCode;
        this.name = name;
    }

    /*
     * Public Instance Methods
     */

    @Override
    public boolean execute(@NotNull PsiElement match, @NotNull ResolveState state) {
        boolean keepProcessing = true;

        if (match instanceof Named) {
            keepProcessing = execute((Named) match, state);
        }

        return keepProcessing;
    }

    @Nullable
    public List<ResolveResult> getResolveResultList() {
        return resolveResultList;
    }

    /*
     * Protected Instance Methods
     */

    /**
     * Decides whether {@code match} matches the criteria being searched for.  All other {@link #execute} methods
     * eventually end here.
     *
     * @return {@code true} to keep processing; {@code false} to stop processing.
     */
    @Override
    protected boolean executeOnAliasedName(@NotNull PsiNamedElement match,
                                           @NotNull String aliasedName,
                                           @NotNull ResolveState state) {
        if (aliasedName.equals(name)) {
            List<String> namePartList = split(name);

            // adds `Foo.SSH` in `alias Foo.SSH`
            addToResolveResultList(match, true);

            // adds `defmodule Foo.SSH` for `alias Foo.SSH`
            addUnaliasedNamedElementsToResolveResultList(match, namePartList);
        } else {
            List<String> namePartList = split(name);
            String firstNamePart = namePartList.get(0);

            // alias Foo.SSH, then SSH.Key is name
            if (aliasedName.equals(firstNamePart)) {
                addToResolveResultList(match, true);

                addUnaliasedNamedElementsToResolveResultList(match, namePartList);
            } else if (incompleteCode && aliasedName.startsWith(name)) {
                addToResolveResultList(match, false);
            }
        }

        return org.elixir_lang.psi.scope.MultiResolve.keepProcessing(incompleteCode, resolveResultList);
    }

    /*
     * Private Instance Methods
     */

    private void addToResolveResultList(@NotNull PsiElement element, boolean validResult) {
        resolveResultList = org.elixir_lang.psi.scope.MultiResolve.addToResolveResultList(
                resolveResultList, new PsiElementResolveResult(element, validResult)
        );
    }

    private void addUnaliasedNamedElementsToResolveResultList(@NotNull PsiNamedElement match, List<String> namePartList) {
        String unaliasedName = unaliasedName(match, namePartList);
        Collection<NamedElement> namedElementCollection = indexedNamedElements(match, unaliasedName);

        for (PsiElement element : namedElementCollection) {
            addToResolveResultList(element, true);
        }
    }
}

File: src/org/elixir_lang/psi/scope/module/MultiResolve.kt
package org.elixir_lang.psi.scope.module

import com.intellij.openapi.project.DumbService
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.ResolveState
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.util.PsiTreeUtil
import org.elixir_lang.Module.concat
import org.elixir_lang.Module.split
import org.elixir_lang.psi.*
import org.elixir_lang.psi.call.Named
import org.elixir_lang.psi.impl.ElixirPsiImplUtil.ENTRANCE
import org.elixir_lang.psi.impl.call.finalArguments
import org.elixir_lang.psi.impl.stripAccessExpression
import org.elixir_lang.psi.scope.Module
import org.elixir_lang.psi.scope.ResolveResultOrderedSet
import org.elixir_lang.psi.scope.VisitedElementSetResolveResult
import org.elixir_lang.psi.scope.maxScope
import org.elixir_lang.psi.stub.index.ModularName
import org.elixir_lang.reference.module.UnaliasedName
import org.elixir_lang.reference.resolver.narrowedScope

class MultiResolve internal constructor(private val name: String, private val incompleteCode: Boolean) : Module() {
    /**
     * Decides whether `match` matches the criteria being searched for.  All other [.execute] methods
     * eventually end here.
     *
     * @return `true` to keep processing; `false` to stop processing.
     */
    override fun executeOnAliasedName(match: PsiNamedElement,
                                      aliasedName: String,
                                      state: ResolveState): Boolean =
        executeOnAliasedName(match, aliasedName, name, state)

    private fun executeOnAliasedName(match: PsiNamedElement, aliasedName: String, targetName: String, state: ResolveState): Boolean {
        val aliasedNameState = state.putVisitedElement(match)

        if (aliasedName == targetName) {
            val namePartList = split(targetName)

            // ALIAS_CALL is only set for `as:` usages
            val aliasCall = state.get(ALIAS_CALL)
            // REQUIRE_CALL is only set for `as:` usages
            val requireCall = state.get(REQUIRE_CALL)

            if (aliasCall == null && requireCall == null) {
                // adds `defmodule Foo.SSH` for `alias Foo.SSH`
                addUnaliasedNamedElementsToResolveResultList(match, namePartList, aliasedNameState.visitedElementSet())
            } else if (aliasCall != null) {
                // adds `Foo.SSH` and `defmodule Foo.SSH` for `alias Foo.SSH, as FSSH`
                aliasCall
                        .finalArguments()!!
                        .first()
                        .stripAccessExpression()
                        .let { it as PsiNamedElement }
                        .let { first ->
                            UnaliasedName.unaliasedName(first)?.let { unaliasedName ->
                                val aliasedNameWithoutAs = split(unaliasedName).last()
                                executeOnAliasedName(
                                        first,
                                        aliasedNameWithoutAs,
                                        aliasedNameWithoutAs,
                                        aliasedNameState.put(ALIAS_CALL, null)
                                )
                            }
                        }
            } else {
                resolveResultOrderedSet.add(match, requireCall.text, true, state.visitedElementSet())
            }
        } else {
            val namePartList = split(name)
            val firstNamePart = namePartList[0]

            // alias Foo.SSH, then SSH.Key is name
            if (aliasedName == firstNamePart) {
                addUnaliasedNamedElementsToResolveResultList(match, namePartList, aliasedNameState.visitedElementSet())
            } else if (aliasedName.startsWith(name)) {
                resolveResultOrderedSet.add(match, "alias ${match.text}", false, state.visitedElementSet())
            }
        }

        return resolveResultOrderedSet.keepProcessing(incompleteCode)
    }

    override fun executeOnModularName(modular: Named, modularName: String, state: ResolveState): Boolean {
        if (modularName.startsWith(name)) {
            val validResult = modularName == name
            resolveResultOrderedSet.add(modular, modularName, validResult, state.visitedElementSet())
        }

        return resolveResultOrderedSet.keepProcessing(incompleteCode)
    }

    fun resolveResults(): List<VisitedElementSetResolveResult> = resolveResultOrderedSet.toList()

    private val resolveResultOrderedSet = ResolveResultOrderedSet()

    private fun addUnaliasedNamedElementsToResolveResultList(match: PsiNamedElement,
                                                             namePartList: List<String>,
                                                             visitedElementSet: Set<PsiElement>) {
        val unaliasedName = unaliasedName(match, namePartList)

        val project = match.project

        if (!DumbService.isDumb(project)) {
            var found = false

            val searchScope = narrowedScope(match, project)

            StubIndex
                    .getInstance()
                    .processElements(
                            ModularName.KEY,
                            unaliasedName,
                            project,
                            searchScope,
                            NamedElement::class.java) {
                resolveResultOrderedSet.add(it, unaliasedName, true, visitedElementSet)
                found = true

                true
            }

            // Transitive alias fallback: if stub lookup found nothing and match is a QualifiableAlias,
            // resolve it through the module scope to follow alias chains
            // (e.g., `alias MyNamespace.Referenced; alias Referenced, as: Refd` — resolving `Refd` needs
            // to follow Referenced → MyNamespace.Referenced transitively)
            if (!found && match is QualifiableAlias) {
                val transitiveResults = resolveResults(match.fullyQualifiedName(), incompleteCode, match)
                for (result in transitiveResults) {
                    val resolvedElement = result.element
                    if (resolvedElement is NamedElement) {
                        resolveResultOrderedSet.add(
                            resolvedElement,
                            resolvedElement.name ?: unaliasedName,
                            result.isValidResult,
                            visitedElementSet + result.visitedElementSet
                        )
                    }
                }
            }
        }
    }

    companion object {
        fun resolveResults(name: String,
                           incompleteCode: Boolean,
                           entrance: PsiElement): List<VisitedElementSetResolveResult> =
                resolveResults(name, incompleteCode, entrance, ResolveState.initial())

        private fun resolveResults(name: String,
                                   incompleteCode: Boolean,
                                   entrance: PsiElement,
                                   state: ResolveState): List<VisitedElementSetResolveResult> {
            val multiResolve = MultiResolve(name, incompleteCode)
            val maxScope = maxScope(entrance)

            PsiTreeUtil.treeWalkUp(
                    multiResolve,
                    entrance,
                    maxScope,
                    state.put(ENTRANCE, entrance).putInitialVisitedElement(entrance)
            )

            return multiResolve.resolveResults()
        }

        private fun unaliasedName(match: PsiNamedElement, namePartList: List<String>): String {
            val matchUnaliasedName = UnaliasedName.unaliasedName(match)

            val unaliasedNamePartList = ArrayList<String>(namePartList.size)
            unaliasedNamePartList.add(matchUnaliasedName!!)

            for (i in 1 until namePartList.size) {
                unaliasedNamePartList.add(namePartList[i])
            }

            return concat(unaliasedNamePartList)
        }
    }
}
